### PR TITLE
Revamp `Num`, `Semigroup`, and `Monoid`

### PR DESCRIPTION
**Pull request by pthariensflame**
They are now proper traits that use `Self`, rather than
explicitly-passed dictionaries.

This is needed for parts of [`Aquifer`](https://github.com/pthariensflame/Aquifer) to work, and in general is a much cleaner interface.  The other typeclasses were not targeted in this pull request, but should probably be gone over similarly as well.

### CODE REVIEW

**Comment by CodaFi** [comment]
My god, this is beautiful.   Absolutely beautiful.  :+1: 

2 things:

- Type annotation, colon, and identifier are separated with spaces.  -At least, that's how I like to see it, and Idris throws a fit when you use colons like that....
- Can you change the Semigroup and Monoid examples in the README?

File: Swiftz/Monoid.swift [hunk 1]
```diff
@@ -9,48 +9,111 @@
...
+        if value().isJust() {
+            return self
+        } else {
+            return other
+        }
+    }
+}
+
+/// The right-biased Maybe Monoid.
+public struct Last<A: Comparable> : Semigroup {
```
Comment (CodaFi) [inline Swiftz/Monoid.swift]: This should be a `Monoid`, no?

Comment (pthariensflame) [inline Swiftz/Monoid.swift]: Yes.  I don't know how I missed that.  :O

**Comment by pthariensflame** [comment]
How does that look?  @typelift/fp

**Comment by CodaFi** [comment]
"All of the formatting" appears to be in spaces and not tabs :P

**Comment by pthariensflame** [comment]
:scream:

**Comment by CodaFi** [comment]
Dankë.  I will tag version 0.1.1 shortly after the merge goes through.

:sail:

**Comment by CodaFi** [comment]
Tagged. Thanks again for all of this.  I've been wanting to redo Num for a while, and this is like reading my mind :sparkles:

**Comment by pthariensflame** [comment]
You're very welcome!  It took the impetus of seeing how ugly explicit dictionary passing actually was to get me to do it.  :P

**Pull request merged**